### DISCUSSION TITLE
Write my code for me please - Battery Level

### DISCUSSION THREAD

**Discussion by YouCanNotBeSerious** [HomeSpan/HomeSpan#727] [Q&A]
Well...embarrassingly and per the title I am basically asking that thing we don't like to be asked - "can you write my code for me" :-(

I have to ask - am beyond frustrated being so close to having my project complete.  Thing is, I have no formal (or practical for that matter) knowledge of OO principles, so while I can hack together things from examples of procedural code most of the time, I am really struggling with this one and hoping a kind soul will help..  OK, enough pity party, if someone was feeling charitable, all I am trying to do is add battery level to my project.

I have (all working) multiple spanpoint contact sensors which are in deep sleep until woken by a reed switch event, when they update the hub with the current state then go back to sleep.  They will be battery powered so naturally I want to send the battery level - I am capable of getting that level via ADC and doing some averaging or whatever I find necessary in code, but I can not work out how to add the battery level to each sensor as a characteristic.

If anyone could help, just a hardcoded battery level is fine, it's the HomeKit and object oriented constructors and such that I don't understand.

Thanks....

Main unit code, works for contact sensor status, but not battery:

```

#include "HomeSpan.h"

struct RemoteDoorSensor : Service::ContactSensor { // this is the constructor for each remote device

  SpanCharacteristic *state;
  SpanCharacteristic *battlevel;
  SpanPoint *remoteDoorSensor;
  const char *name;
  int doorstate;
  int level;

  RemoteDoorSensor(const char *name, const char*macAddress) : Service::ContactSensor(){ 
    this->name=name;  
    state=new Characteristic::ContactSensorState(0);      // set initial state
    battlevel=new Characteristic::BatteryLevel(50);
    remoteDoorSensor=new SpanPoint(macAddress,0,sizeof(doorstate));   

  } // end constructor

  void loop(){  // confusing to have 2 loop() functions! I need more OO knowledge, this loop seems to be for the constructor, main loop is below.

    if(remoteDoorSensor->get(&doorstate))
       {
          state->setVal(doorstate);
          battlevel->setVal(level);
          LOG1("batt= %i", level);
          LOG1("Sensor %s update: STATE=%i (%s) \n",name, doorstate, doorstate?"OPEN":"CLOSED");
       }
    } 
};

//////////////////////////////////////

void setup() {
  Serial.begin(115200);
  Serial.println("This hub's MAC is: " + WiFi.macAddress());

  homeSpan.setStatusPin(5);
  homeSpan.setControlPin(16); 
  homeSpan.setLogLevel(1);
  homeSpan.enableOTA();        

  homeSpan.begin(Category::Bridges,"Sensor Hub");    // the hub
  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 

  new SpanAccessory();                               // first remote device                   
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Contact Sensor 1");  
    new RemoteDoorSensor("Door 1","68:67:25:B6:10:24");       

  new SpanAccessory();                               // second remote device ... testing battery level                  
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Contact Sensor 2");  
  new RemoteDoorSensor("Door 2","68:67:25:B7:EA:74");
    new Service::BatteryService();
       new Characteristic::BatteryLevel();    // a sub characteristic of this accessory??
       
} 

//////////////////////////////////////

void loop()
{
  homeSpan.poll();    
} 

//////////////////////////////////////

```

Remote unit code, works for contact sensor status, but not battery:

```

#include "HomeSpan.h"

#define WAKEUP_PIN_OPEN 4   
#define WAKEUP_PIN_CLOSED 1  // reed switch has NO and NC contacts

RTC_DATA_ATTR int state = 0;  //0 = closed, 1 = open, RTC variable is retained over sleep 
SpanPoint *mainDevice;
int wakeup_reason = esp_sleep_get_wakeup_cause();

void setup() 
{
  Serial.begin(115200);
  Serial.println("Starting...reset reason is: " + String(wakeup_reason));  //7 means a pin woke us from deep sleep. might use this to stay in OTA mode if woken by poweron or reset button );  
  Serial.println("This remote device's MAC is: " + WiFi.macAddress());   
 
  pinMode(WAKEUP_PIN_CLOSED, INPUT_PULLUP);   // this sets pullups in the normal way, but while asleep only the one selected as a wake source stays as a pullup.
  pinMode(WAKEUP_PIN_OPEN, INPUT_PULLUP); 
  
  mainDevice=new SpanPoint("24:DC:C3:A0:1F:F4",sizeof(state),0);    
}

void loop() 
{

// NOTE on a wakeup, we swap the wakeup source for next time i.e. if woken by OPEN we go to sleep and wait for CLOSED. 
// We need to do this otherwise once woken by a closed to open transition, we would continually be re-awoken for as long as the door is open and vice versa.
// If we get out of sync, an OPEN/CLOSE cycle gets us back

if (state == 0) // we were CLOSED before we went to sleep, so must have been woken by an OPEN
{
  state = 1;
  Serial.println("OPEN - going to next wake on CLOSED");
  esp_deep_sleep_enable_gpio_wakeup(1 << WAKEUP_PIN_CLOSED, ESP_GPIO_WAKEUP_GPIO_LOW);  // set up next wakeup
}
else 
{
  state = 0;
  Serial.println("CLOSED - going to next wake on OPEN");
  esp_deep_sleep_enable_gpio_wakeup(1 << WAKEUP_PIN_OPEN, ESP_GPIO_WAKEUP_GPIO_LOW);  // set up next wakeup
}
  boolean success = mainDevice->send(&state);                 // this will show as success as long as the MAIN DEVICE is running 
  Serial.printf("Send %s\n",success?"Succeded":"Failed");
  Serial.println("State = " + String(state));
  Serial.println("Going to sleep in 250ms...");
  //Serial.flush();   //good when debugging, but causes unit to wait for the serial port to acknowledge so when not connected to a port we wait here forever
  delay(250);  // debouncing....works for now, look at span button though 
  esp_deep_sleep_start();  // night night
}


```

**Comment by frankonski** [comment]
Hi there! Unfortunately, I won't be writing code for you (no hard feelings!), but will point you to some docs, which will hopefully get you back on track.

Looking at the [Service List](https://github.com/HomeSpan/HomeSpan/blob/master/docs/ServiceList.md#service-list) and their optional characteristics, what we can see is that a **ContactSensor** cannot have a _BatteryLevel_ characteristic, but only a _StatusLowBattery_ characteristic. When HomeSpan starts, its prints its accessory database along with Errors and Warnings. I betcha there is a Warning when you code runs about something like "ContactSensor does not support BatteryLevel characteristic".

In your code above, see `struct RemoteDoorSensor` as your own customized version of a `ContactSensor` service. That service cannot have a _BatteryLevelService_. However, if you look in Home app, or Eve, and look at the details of your _Contact Sensor 2_, I believe you should see the battery level for it, as you are defining this second accessory with two services: `RemoteDoorSensor` (which is a ContactSensor) and `BatteryService`.

As for your question around `loop()`, read the doc of `virtual void loop()` in the [Reference](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Reference.md), that should explain why you have to define it.

**Reply by YouCanNotBeSerious** [reply]
Hehehe thanks for the reply, I'll admit there is some frustration at my own lack of education going on here :-)

I do appreciate your pointers, sometimes a step back is needed (and some thisguy_deep_sleep()!) and I think i will RTFM sstarting with your information before trying again as I did sort of dive straight in and i think i piled misunderstandings together.

**Reply by HomeSpan** [reply]
You may also want to look at the Battery Service Example under HomeSpan Reference Sketches.

**Comment by YouCanNotBeSerious** [comment]
@frankonski your pointers helped a lot, **I now have individual battery low status (not level) per remote sensor.**. I added the characteristic under RemoteDoorSensor.

>Looking at the [Service List](https://github.com/HomeSpan/HomeSpan/blob/master/docs/ServiceList.md#service-list) and their optional characteristics, what we can see is that a ContactSensor cannot have a BatteryLevel characteristic, but only a StatusLowBattery characteristic. When HomeSpan starts, its prints its accessory database along with Errors and Warnings. I betcha there is a Warning when you code runs about something like "ContactSensor does not support BatteryLevel characteristic".

Of course the ContactSensor only has StatsuBatteryLow, and yes upon startup if I scroll back in what I now know is my too small serial monitor buffer I now see the appropriate warnings. Checking for "errors =0 and warnings = 0" every time now!

> if you look in Home app, or Eve, and look at the details of your Contact Sensor 2, I believe you should see the battery level for it, as you are defining this second accessory with two services: RemoteDoorSensor (which is a ContactSensor) and BatteryService.

Indeed I do! That alone feels like I've achieved something :-) But that comes with warnings from homespan and seems similar to [this](https://github.cwom/HomeSpan/HomeSpan/discussions/442) and [this.](https://github.com/HomeSpan/HomeSpan/discussions/156)

So adding the StatusLowBattery to the contact sensor worked easily enough, but I now need to get my head around the battery service - as in, how does that get associated to each contact sensor rather than be a seperate service.  The issues linked above are what I am reading now on that. I believe I would make a new struct for the battery service where it can be defined and then add it to 

```
new SpanAccessory();                               // first remote device                   
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Contact Sensor 1");  
      new RemoteDoorSensor("Door 1","68:67:25:B6:10:24");       
     new myNewBatteryStruct ().  <--- here, so the battery service is part of the contact sensor accessory?
```

> @HomeSpan You may also want to look at the Battery Service Example under HomeSpan Reference Sketches.

Seems unbelievable, but I had not seen this set of sketches, even though they are clearly noted. There's just so much to learn!  For now I have my very basic read ADC and correlate voltage to "low" or not, but I will look to improve it using some of the things in that example, or just use that class verbatim.  All of HomeSpan as a project amazes me somewhat still with the amount of work you have done and made available to us for free, I'm drinking from the firehose here.
